pxf/pxf-json/src/main/java/org/apache/hawq/pxf/plugins/json/JsonResolver.java

```diff
@@ -0,0 +1,259 @@
+ * and JSON array indexing.
+ */
+public class JsonResolver extends Plugin implements ReadResolver {
+
+	private ArrayList<OneField> list = new ArrayList<OneField>();
+
+	public JsonResolver(InputData inputData) throws Exception {
+		super(inputData);
+	}
+
+	@Override
+	public List<OneField> getFields(OneRow row) throws Exception {
+		list.clear();
+
+		// key is a Text object
+		JsonNode root = JsonInputFormat.decodeLineToJsonNode(row.getKey().toString());
+
+		// if we weren't given a null object
+		if (root != null) {
+			// Iterate through the column definition and fetch our JSON data
+			for (int i = 0; i < inputData.getColumns(); ++i) {
+
+				// Get the current column description
+				ColumnDescriptor cd = inputData.getColumn(i);
+				DataType columnType = DataType.get(cd.columnTypeCode());
+
+				// Get the JSON projections from the column name
+				// For example, "user.name" turns into ["user","name"]
+				String[] projs = cd.columnName().split("\\.");
+
+				// Move down the JSON path to the final name
+				JsonNode node = getPriorJsonNode(root, projs);
+
+				// If this column is an array index, ex. "tweet.hashtags[0]"
+				if (isArrayIndex(projs)) {
+
+					// Get the node name and index
+					String nodeName = getArrayName(projs);
+					int arrayIndex = getArrayIndex(projs);
+
+					// Move to the array node
+					node = node.get(nodeName);
+
+					// If this node is null or missing, add a null value here
+					if (node == null || node.isMissingNode()) {
+						addNullField(columnType);
+					} else if (node.isArray()) {
+						// If the JSON node is an array, then add it to our list
+						addFieldFromJsonArray(columnType, node, arrayIndex);
+					} else {
+						throw new InvalidParameterException(nodeName + " is not an array node");
+					}
+				} else {
+					// This column is not an array type
+					// Move to the final node
+					node = node.get(projs[projs.length - 1]);
+
+					// If this node is null or missing, add a null value here
+					if (node == null || node.isMissingNode()) {
+						addNullField(columnType);
+					} else {
+						// Else, add the value to the record
+						addFieldFromJsonNode(columnType, node);
+					}
+				}
+			}
+		}
+
+		return list;
+	}
+
+	/**
+	 * Iterates down the root node to the prior JSON node. This node is used
+	 * 
+	 * @param root
+	 * @param projs
+	 * @return
+	 */
+	private JsonNode getPriorJsonNode(JsonNode root, String[] projs) {
+
+		// Iterate through all the tokens to the desired JSON node
+		JsonNode node = root;
+		for (int j = 0; j < projs.length - 1; ++j) {
+			node = node.path(projs[j]);
+		}
+
+		return node;
+	}
+
+	/**
+	 * Gets a boolean value indicating if this column is an array index column
+	 * 
+	 * @param projs
+	 *            The array of JSON projections
+	 * @throws ArrayIndexOutOfBoundsException
+	 */
+	private boolean isArrayIndex(String[] projs) {
+		return projs[projs.length - 1].contains("[") && projs[projs.length - 1].contains("]");
```

Comment:
will chanage the array name/index retrieval to to use the Pattern.compile("(.*)\\[([0-9]+)\\]") matching groups instead